core/server/common/src/main/java/alluxio/master/PrimarySelectorClient.java

```diff
@@ -132,7 +139,54 @@ public void start(InetSocketAddress address) throws IOException {
 
   @Override
   public void stateChanged(CuratorFramework client, ConnectionState newState) {
-    setState(State.SECONDARY);
+    // Handle state change.
+    switch (newState) {
+      case CONNECTED:
+      case LOST:
+        setState(State.SECONDARY);
+        break;
+      case SUSPENDED:
+        if (mConnectionErrorPolicy == ZookeeperConnectionErrorPolicy.STANDARD) {
+          // Assume SECONDARY mode under standard policy.
+          setState(State.SECONDARY);
+        }
+        break;
+      case RECONNECTED:
+        if (mConnectionErrorPolicy == ZookeeperConnectionErrorPolicy.STANDARD) {
+          // Assume SECONDARY mode under standard policy.
+          setState(State.SECONDARY);
+        } else {
+          // Try to retain existing PRIMARY role under session policy.
+          if (getState() == State.PRIMARY) {
+            /**
+             * Do a sanity check when reconnected for a selector with "PRIMARY" mode.
+             * This is to ensure that curator reconnected with the same Id.
+             * Hence, guaranteeing Zookeeper state for this instance was preserved.
+             */
+            try {
+              long reconnectSessionId = client.getZookeeperClient().getZooKeeper().getSessionId();
+              if (mLeaderZkSessionId != reconnectSessionId) {
+                LOG.warn(String.format(
+                    "Curator reconnected under a different session. "
+                        + "Old sessionId: %x, New sessionId: %x",
```

Comment:
are all session ids typically printed in hex?

Reply:
Yes. We not printing in hex was creating inconsistency between zk traces.